# Clustering Neighborhoods in Toronto

#### Author: David Rodriguez Castro

In this assigment will going to explore, segment, and cluster the neighborhoods in the city of Toronto based on the postal code and borough information

This notebook includes Questions 1, 2 and 3 of the Assignment. They were separated by Section Headers

In [1]:
# <-- Read it into a pandas  dataframe to structure the  format --> 
import pandas as pd


# Question 1

## Import Data

On this part we need to know is that we are able to make the connection to the page and 
verify that the data is completed and have the correct information to perform the assignment

In [2]:
import requests

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

In [4]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [5]:
len(wiki_data), type(wiki_data)

(3, list)

We need the first table alone, so dropping the other tables

In [6]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

Dropping Borough which are not assigned

In [7]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Grouping the records based on Postal Code

In [8]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [9]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [10]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [11]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [12]:
df.shape

(103, 3)

In [53]:
print("The number of rows and colums are:", df.shape)

The number of rows and colums are: (103, 3)


# Question 2

## Installing geocoder

Now, we want to the description of the  locations, postal codes, Borough,Neighborhood
and geographic coordinates(latitude/longitude) to identify the correct location on the map

In [13]:
pip install geocoder

     |████████████████████████████████| 98 kB 6.9 MB/s eta 0:00:011
Note: you may need to restart the kernel to use updated packages.


In [14]:
import geocoder # import geocoder

I decide to proceed import the CSV file from the URL

In [15]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

Alternative method to initiate geocoder

In [ ]:
#-->initialize your variable to None
#import geocoder # import geocoder

# initialize your variable to None
#lat_lng_coords = None

# loop until you get the coordinates
#while(lat_lng_coords is None):
#  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
#  lat_lng_coords = g.latlng

#latitude = lat_lng_coords[0]
#longitude = lat_lng_coords[1] 

In [16]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Since the dimensions are the same, we can try to join on the postal codes to get the required data.

Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on it

In [17]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [18]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [19]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [20]:
combined_data.shape

(103, 5)

Solution: The result of combined data shows 103 rows and 5 columns as expected.

# Question 3

Using  Foursquare API for visualizing the clusters classified by different venue categories on the Toronto neighborhood cities.We need to define functions to get all the venue categories in Toronto

In [21]:
from geopy.geocoders import Nominatim

In [22]:
pip install folium

     |████████████████████████████████| 94 kB 4.2 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [23]:
import folium

In [24]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


### View Clustering Neighborhoods in Toronto Maps from my Jupiter Notebook
[Click here to see the Maps](https://eu-gb.dataplatform.cloud.ibm.com/analytics/notebooks/v2/7b393eaf-fd25-4538-a2d6-8ce51b075c07/view?access_token=105453fa0258515889b43fbeeb4ab2438d86e5eda3bda306abc487eb5d7b34e7)

In [25]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [26]:
CLIENT_ID = '1BF4RHWZ32K1DEJYNHZM2JGXB0M2L2RAAHIDCEEDVLHPDADM' 
CLIENT_SECRET = 'JDBARAF1QRQMGB0SNMJM3INWS2CFCAKYASBVRCIPPOT0XQPZ'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 1BF4RHWZ32K1DEJYNHZM2JGXB0M2L2RAAHIDCEEDVLHPDADM
CLIENT_SECRET:JDBARAF1QRQMGB0SNMJM3INWS2CFCAKYASBVRCIPPOT0XQPZ


Create a function to get all the venue categories in Toronto

In [27]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood

In [28]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [29]:
venues_in_toronto.shape

(1338, 5)

In [60]:
print("The Result shows:",venues_in_toronto.shape,'(Records,Colums)')

The Result shows: (1338, 5) (Records,Colums)


In [30]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3         Parkwoods               43.753259               -79.329656   
4  Victoria Village               43.725882               -79.315572   

                               Venue              Venue Category  
0                    Brookbanks Park                        Park  
1                Towns On The Ravine                       Hotel  
2                      Variety Store           Food & Drink Shop  
3  Corrosion Service Company Limited  Construction & Landscaping  
4             Victoria Village Arena                Hockey Arena

Checking the Venues based on Neighbourhood

In [31]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                             Parkwoods   
3                                             Parkwoods   
4                                      Victoria Village   
...                                                 ...   
1323  Mimico NW, The Queensway West, South of Bloor,...   
1324  Mimico NW, The Queensway West, South of Bloor,...   
1325  Mimico NW, The Queensway West, South of Bloor,...   
1326  Mimico NW, The Queensway West, South of Bloor,...   
1327  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude  \
0                  43.753259               -79.329656   
1                  43.753259               -79.329656   
2                  43.753259               -79.329656   
3                  43.753259               -79.329656   
4                  43.725882               -79.315572   
...                      ...                      ...   
1323               43.628841               -79.520999   
1324               43.628841               -79.520999   
1325               43.628841               -79.520999   
1326               43.628841               -79.520999   
1327               43.628841               -79.520999   

                                  Venue              Venue Category  
0                       Brookbanks Park                        Park  
1                   Towns On The Ravine                       Hotel  
2                         Variety Store           Food & Drink Shop  
3     Corrosion Service Company Limited  Construction & Landscaping  
4                Victoria Village Arena                Hockey Arena  
...                                 ...                         ...  
1323                         Wingporium                 Wings Joint  
1324                   South St. Burger                Burger Joint  
1325                          Dollarama              Discount Store  
1326                     Healthy Planet             Supplement Shop  
1327                           7-Eleven           Convenience Store  

[416 rows x 5 columns]

In [59]:
print('The results indicate 416 records for each neighbourhood')

The results indicate 416 records for each neighbourhood


In [32]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
Airport Service     CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                     Little Portugal, Trinity   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                        Lawrence Manor, Lawrence Heights   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
Airport Service                  43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.400049   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.733283               -79.419750   
Yoga Studio                      43.715383               -79.321558   

                                               Venue  
Venue Category                                        
Accessories Store                Ardene Shoes Outlet  
Airport              Toronto Downsview Airport (YZD)  
Airport Food Court                 Billy Bishop Café  
Airport Lounge                         Porter Lounge  
Airport Service     Porter Airlines Check-In Counter  
...                                              ...  
Warehouse Store                               Costco  
Wine Bar                             Paris Paris Bar  
Wings Joint                               Wingporium  
Women's Store                          Want Boutique  
Yoga Studio                                YogaSpace  

[235 rows x 4 columns]

In [61]:
print("There are around 235 different types of Venue Categories")

There are around 235 different types of Venue Categories


In [33]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Lounge  \
0                     0        0                   0               0   
1                     0        0                   0               0   
2                     0        0                   0               0   
3                     0        0                   0               0   
4                     0        0                   0               0   
...                 ...      ...                 ...             ...   
1333                  0        0                   0               0   
1334                  0        0                   0               0   
1335                  0        0                   0               0   
1336                  0        0                   0               0   
1337                  0        0                   0               0   

      Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                   0                 0                    0             0   
1                   0                 0                    0             0   
2                   0                 0                    0             0   
3                   0                 0                    0             0   
4                   0                 0                    0             0   
...               ...               ...                  ...           ...   
1333                0                 0                    0             0   
1334                0                 0                    0             0   
1335                0                 0                    0             0   
1336                0                 0                    0             0   
1337                0                 0                    0             0   

      Aquarium  Art Gallery  ...  Turkish Restaurant  \
0            0            0  ...                   0   
1            0            0  ...                   0   
2            0            0  ...                   0   
3            0            0  ...                   0   
4            0            0  ...                   0   
...        ...          ...  ...                 ...   
1333         0            0  ...                   0   
1334         0            0  ...                   0   
1335         0            0  ...                   0   
1336         0            0  ...                   0   
1337         0            0  ...                   0   

      Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                                 0                 0            0   
1                                 0                 0            0   
2                                 0                 0            0   
3                                 0                 0            0   
4                                 0                 0            0   
...                             ...               ...          ...   
1333                              0                 0            0   
1334                              0                 0            0   
1335                              0                 0            0   
1336                              0                 0            0   
1337                              0                 0            0   

      Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                         0                0         0            0   
1                         0                0         0            0   
2                         0                0         0            0   
3                         0                0         0            0   
4                         0                0         0            0   
...                     ...              ...       ...          ...   
1333                      0                0         0            0   
1334                      0                0         0            0   
1335                      0                0       

Adding the neighbourhood to the encoded dataframe

In [34]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2         Parkwoods                  0        0                   0   
3         Parkwoods                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Turkish Restaurant  \
0             0         0  ...                   0   
1             0         0  ...                   0   
2             0         0  ...                   0   
3             0         0  ...                   0   
4             0         0  ...                   0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 236 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood

In [35]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Lounge  Airport Service  \
0      0.0                 0.0             0.0              0.0   
1      0.0                 0.0             0.0              0.0   
2      0.0                 0.0             0.0              0.0   
3      0.0                 0.0             0.0              0.0   
4      0.0                 0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  ...  \
0               0.0             0.000000           0.0       0.0  ...   
1               0.0             0.000000           0.0       0.0  ...   
2               0.0             0.000000           0.0       0.0  ...   
3               0.0             0.000000           0.0       0.0  ...   
4               0.0             0.038462           0.0       0.0  ...   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                 0.0                            0.0               0.0   
1                 0.0                            0.0               0.0   
2                 0.0                            0.0               0.0   
3                 0.0                            0.0               0.0   
4                 0.0                            0.0               0.0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0          0.0                    0.0              0.0       0.0          0.0   
1          0.0                    0.0              0.0       0.0          0.0   
2          0.0                    0.0              0.0       0.0          0.0   
3          0.0                    0.0              0.0       0.0          0.0   
4          0.0                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0       0.000000          0.0  
1       0.000000          0.0  
2       0.000000          0.0  
3       0.000000          0.0  
4       0.038462          0.0  

[5 rows x 236 columns]

Create function to get the top most common venue categories

In [36]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [37]:
import numpy as np

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods

In [38]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                Coffee Shop   
3                                  Bayview Village                       Café   
4                Bedford Park, Lawrence Manor East             Sandwich Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot        Clothing Store          Skating Rink   
1              Pharmacy        Sandwich Place                  Pool   
2                  Bank                 Diner   Fried Chicken Joint   
3    Chinese Restaurant                  Bank   Japanese Restaurant   
4    Italian Restaurant           Coffee Shop     Indian Restaurant   

  5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0                Lounge                Yoga Studio          Dessert Shop   
1                   Pub                Coffee Shop                   Gym   
2      Sushi Restaurant  Middle Eastern Restaurant        Ice Cream Shop   
3      Department Store                Escape Room     Electronics Store   
4                   Pub                       Café               Butcher   

         8th Most Common Venue        9th Most Common Venue  \
0            Electronics Store  Eastern European Restaurant   
1                      Gay Bar              Coworking Space   
2                   Restaurant            Mobile Phone Shop   
3  Eastern European Restaurant                    Drugstore   
4             Sushi Restaurant                 Liquor Store   

  10th Most Common Venue  
0              Drugstore  
1         Discount Store  
2          Deli / Bodega  
3             Donut Shop  
4             Hobby Shop

In [39]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

Make the model to cluster our Neighbourhoods

In [40]:
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [41]:
kmeans.labels_[0:100]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 4, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       4, 1, 4, 1, 1, 2, 4, 1, 4, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4,
       1, 1, 1, 1, 3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1,
       4, 1, 1, 1, 1, 1, 4], dtype=int32)

Add the clustering Label column to the top 10 common venue categories

In [42]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [43]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels   1st Most Common Venue  \
0  43.753259 -79.329656             1.0                    Park   
1  43.725882 -79.315572             1.0             Pizza Place   
2  43.654260 -79.360636             1.0             Coffee Shop   
3  43.718518 -79.464763             1.0  Furniture / Home Store   
4  43.662301 -79.389494             1.0             Coffee Shop   

        2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0  Construction & Landscaping      Food & Drink Shop                 Hotel   
1                 Coffee Shop  Portuguese Restaurant     French Restaurant   
2                        Park                 Bakery        Breakfast Spot   
3              Clothing Store      Accessories Store           Coffee Shop   
4           College Cafeteria                  Diner      Sushi Restaurant   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0           Yoga Studio         Deli / Bodega  Eastern European Restaurant   
1          Hockey Arena                 Diner                Deli / Bodega   
2                  Café               Theater                  Yoga Studio   
3    Miscellaneous Shop              Boutique                  Event Space   
4           Yoga Studio                  Café               Sandwich Place   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0              Drugstore            Donut Shop                Dog Run  
1       Department Store          Dessert Shop     Dim Sum Restaurant  
2       Greek Restaurant                   Pub  Performing Arts Venue  
3  Vietnamese Restaurant         Women's Store           Dance Studio  
4          Smoothie Shop   Distribution Center          Burrito Place

In [44]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

In [45]:
import matplotlib.cm as cm
import matplotlib.colors as colors

Plotting the clusters on the map

### View Clustering Neighborhoods in Toronto Maps from my Jupiter Notebook
[Click here to see the Maps](https://eu-gb.dataplatform.cloud.ibm.com/analytics/notebooks/v2/7b393eaf-fd25-4538-a2d6-8ce51b075c07/view?access_token=105453fa0258515889b43fbeeb4ab2438d86e5eda3bda306abc487eb5d7b34e7)

In [46]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Cluster 1

In [47]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             0.0                Bakery           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11           Event Space           Escape Room     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11  Eastern European Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
11               Dog Run    Distribution Center

Cluster 2

In [48]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels   1st Most Common Venue  \
0          North York             1.0                    Park   
1          North York             1.0             Pizza Place   
2    Downtown Toronto             1.0             Coffee Shop   
3          North York             1.0  Furniture / Home Store   
4    Downtown Toronto             1.0             Coffee Shop   
..                ...             ...                     ...   
98          Etobicoke             1.0                   River   
99   Downtown Toronto             1.0        Sushi Restaurant   
100      East Toronto             1.0             Yoga Studio   
101         Etobicoke             1.0          Baseball Field   
102         Etobicoke             1.0             Social Club   

          2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0    Construction & Landscaping      Food & Drink Shop                 Hotel   
1                   Coffee Shop  Portuguese Restaurant     French Restaurant   
2                          Park                 Bakery        Breakfast Spot   
3                Clothing Store      Accessories Store           Coffee Shop   
4             College Cafeteria                  Diner      Sushi Restaurant   
..                          ...                    ...                   ...   
98                   Smoke Shop                   Pool           Curling Ice   
99                  Coffee Shop                   Café              Beer Bar   
100                  Smoke Shop          Auto Workshop               Brewery   
101                Home Service            Yoga Studio      Department Store   
102           Convenience Store         Sandwich Place            Kids Store   

    5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0             Yoga Studio         Deli / Bodega  Eastern European Restaurant   
1            Hockey Arena                 Diner                Deli / Bodega   
2                    Café               Theater                  Yoga Studio   
3      Miscellaneous Shop              Boutique                  Event Space   
4             Yoga Studio                  Café               Sandwich Place   
..                    ...                   ...                          ...   
98              Drugstore            Donut Shop                      Dog Run   
99       Ramen Restaurant   Martial Arts School                   Steakhouse   
100         Burrito Place               Butcher                   Comic Shop   
101           Escape Room     Electronics Store  Eastern European Restaurant   
102  Fast Food Restaurant          Burger Joint                Burrito Place   

     8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Drugstore            Donut Shop                Dog Run  
1         Department Store          Dessert Shop     Dim Sum Restaurant  
2         Greek Restaurant                   Pub  Performing Arts Venue  
3    Vietnamese Restaurant         Women's Store           Dance Studio  
4            Smoothie Shop   Distribution Center          Burrito Place  
..                     ...                   ...                    ...  
98     Distribution Center        Discount Store                  Diner  
99       Indian Restaurant  Ethiopian Restaurant            Escape Room  
100         Farmers Market  Fast Food Restaurant                 Garden  
101              Drugstore            Donut Shop                Dog Run  
102         Discount Store       Supplement Shop                 Bakery  

[87 rows x 12 columns]

Cluster 3

In [49]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             2.0  Fast Food Restaurant           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
6         Deli / Bodega     Electronics Store  Eastern European Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6             Drugstore            Donut Shop               Dog Run   

  9th Most Common Venue 10th Most Common Venue  
6   Distribution Center         Discount Store

Cluster 4

In [50]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
32  Scarborough             3.0            Playground          Dance Studio   

   3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
32     Electronics Store  Eastern European Restaurant             Drugstore   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
32            Donut Shop               Dog Run   Distribution Center   

   9th Most Common Venue 10th Most Common Venue  
32        Discount Store                  Diner

Cluster 5

In [51]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
21              York             4.0                  Park   
35         East York             4.0                  Park   
61   Central Toronto             4.0                  Park   
64              York             4.0                  Park   
66        North York             4.0                  Park   
77         Etobicoke             4.0                  Park   
83   Central Toronto             4.0                  Park   
85       Scarborough             4.0            Playground   
91  Downtown Toronto             4.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21         Women's Store                  Pool           Yoga Studio   
35         Metro Station     Convenience Store           Yoga Studio   
61           Swim School              Bus Line           Yoga Studio   
64           Yoga Studio          Dance Studio     Electronics Store   
66     Convenience Store           Yoga Studio         Deli / Bodega   
77        Sandwich Place     Mobile Phone Shop           Yoga Studio   
83                 Trail            Restaurant           Yoga Studio   
85                  Park          Intersection           Yoga Studio   
91            Playground                 Trail           Yoga Studio   

          5th Most Common Venue        6th Most Common Venue  \
21                 Dance Studio  Eastern European Restaurant   
35                Deli / Bodega  Eastern European Restaurant   
61                Deli / Bodega  Eastern European Restaurant   
64  Eastern European Restaurant                    Drugstore   
66            Electronics Store  Eastern European Restaurant   
77                Deli / Bodega  Eastern European Restaurant   
83                 Dance Studio  Eastern European Restaurant   
85                 Dance Studio  Eastern European Restaurant   
91                 Dance Studio  Eastern European Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
21             Drugstore            Donut Shop               Dog Run   
35             Drugstore            Donut Shop               Dog Run   
61             Drugstore            Donut Shop               Dog Run   
64            Donut Shop               Dog Run   Distribution Center   
66             Drugstore            Donut Shop               Dog Run   
77             Drugstore            Donut Shop               Dog Run   
83             Drugstore            Donut Shop               Dog Run   
85             Drugstore            Donut Shop               Dog Run   
91             Drugstore            Donut Shop               Dog Run   

   10th Most Common Venue  
21    Distribution Center  
35    Distribution Center  
61    Distribution Center  
64         Discount Store  
66    Distribution Center  
77    Distribution Center  
83    Distribution Center  
85    Distribution Center  
91    Distribution Center